core/src/apresolve.rs

```diff
@@ -1,132 +1,124 @@
-        ApResolveData::default()
-    });
-
-    let accesspoint = select_ap(data.accesspoint, AP_FALLBACK, ap_port);
-    let dealer = select_ap(data.dealer, DEALER_FALLBACK, ap_port);
-    let spclient = select_ap(data.spclient, SPCLIENT_FALLBACK, ap_port);
+impl ApResolver {
+    fn split_aps(data: Vec<String>) -> Vec<SocketAddress> {
+        data.into_iter()
+            .filter_map(|ap| {
+                let mut split = ap.rsplitn(2, ':');
+                let port = split
+                    .next()
+                    .expect("rsplitn should not return empty iterator");
+                let host = split.next()?.to_owned();
+                let port: u16 = port.parse().ok()?;
+                Some((host, port))
+            })
+            .collect()
+    }
 
-    AccessPoints {
-        accesspoint,
-        dealer,
-        spclient,
+    fn find_ap(&self, data: &[SocketAddress]) -> usize {
+        match self.session().config().proxy {
+            Some(_) => data
+                .iter()
+                .position(|(_, port)| *port == self.session().config().ap_port.unwrap_or(443))
+                .expect("No access points available with that proxy port."),
+            None => 0, // just pick the first one
+        }
     }
-}
 
-#[cfg(test)]
-mod test {
-    use std::net::ToSocketAddrs;
+    async fn try_apresolve(&self) -> Result<ApResolveData, Box<dyn Error>> {
+        let req = Request::builder()
+            .method("GET")
+            .uri("http://apresolve.spotify.com/?type=accesspoint&type=dealer&type=spclient")
+            .body(Body::empty())
+            .unwrap();
 
-    use super::apresolve;
+        let body = self.session().http_client().request_body(req).await?;
+        let data: ApResolveData = serde_json::from_slice(body.as_ref())?;
 
-    #[tokio::test]
-    async fn test_apresolve() {
-        let aps = apresolve(None, None).await;
+        Ok(data)
+    }
 
-        // Assert that the result contains a valid host and port
-        aps.accesspoint.to_socket_addrs().unwrap().next().unwrap();
-        aps.dealer.to_socket_addrs().unwrap().next().unwrap();
-        aps.spclient.to_socket_addrs().unwrap().next().unwrap();
+    async fn apresolve(&self) {
+        let result = self.try_apresolve().await;
+        self.lock(|inner| {
+            let data = match result {
+                Ok(data) => data,
+                Err(e) => {
+                    warn!("Failed to resolve access points, using fallbacks: {}", e);
+                    ApResolveData::default()
+                }
+            };
+
+            inner.data.accesspoint = Self::split_aps(data.accesspoint);
+            inner.data.dealer = Self::split_aps(data.dealer);
+            inner.data.spclient = Self::split_aps(data.spclient);
+        })
     }
 
-    #[tokio::test]
-    async fn test_apresolve_port_443() {
-        let aps = apresolve(None, Some(443)).await;
+    fn is_empty(&self) -> bool {
```

Comment:
This implementation won't work correctly with a proxy. There still be access points on the list, just not with the chosen proxy port, and that will cause a panic on line 60. It might be better to filter the list in `split_aps`.